python_modules/dagster/dagster/components/core/tree.py

```diff
@@ -0,0 +1,87 @@
+import importlib
+from functools import cached_property
+from pathlib import Path
+from types import ModuleType
+from typing import Optional
+
+from dagster_shared.record import record
+from dagster_shared.serdes.objects.package_entry import json_for_all_components
+from typing_extensions import TypeVar
+
+from dagster._core.definitions.definitions_class import Definitions
+from dagster.components.component.component import Component
+from dagster.components.core.context import ComponentLoadContext, use_component_load_context
+from dagster.components.core.defs_module import DefsFolderComponent
+from dagster.components.core.package_entry import discover_entry_point_package_objects
+from dagster.components.core.snapshot import get_package_entry_snap
+
+PLUGIN_COMPONENT_TYPES_JSON_METADATA_KEY = "plugin_component_types_json"
+
+TComponent = TypeVar("TComponent", bound=Component)
+
+
+@record(
+    checked=False,  # cant handle ModuleType
+)
+class ComponentTree:
+    defs_module: ModuleType
+    project_root: Path
+
+    @staticmethod
+    def find(path: Optional[Path] = None):
+        from dagster_dg.context import DgContext
+
+        # default to cwd
+        path = path or Path()
+
+        # replace with dagster_shared impl of path crawl and config resolution
+        dg_context = DgContext.for_project_environment(path, command_line_config={})
+
+        return ComponentTree(
+            defs_module=importlib.import_module(dg_context.defs_module_name),
+            project_root=dg_context.root_path,
+        )
+
+    @cached_property
+    def load_context(self):
+        return ComponentLoadContext.for_module(
+            defs_module=self.defs_module,
+            project_root=self.project_root,
+        )
+
+    @cached_property
+    def root(self) -> DefsFolderComponent:
+        return DefsFolderComponent.get(self.load_context)
+
+    def load_defs(self) -> Definitions:
+        with use_component_load_context(self.load_context):
+            library_objects = discover_entry_point_package_objects()
+            snaps = [get_package_entry_snap(key, obj) for key, obj in library_objects.items()]
+            components_json = json_for_all_components(snaps)
+
+            return Definitions.merge(
+                self.root.build_defs(self.load_context),
+                Definitions(metadata={PLUGIN_COMPONENT_TYPES_JSON_METADATA_KEY: components_json}),
+            )
```

Comment:
This is duplicated over from `load_defs` as this introduces the requirement that the root is a `DefsFolderComponent` .  Originally I tried to migrate `load_defs` to call this, but there a significant number of tests that would need to be updated and that api is marked `@public` so i am concerned about breaking changes.